clang/tools/checked-c-convert/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -1,127 +1,573 @@
+      std::string funcName = getCalledFunctionName(CE);
+      // check if the called function is a known allocator?
+      return AllocatorSizeAssoc.find(funcName) !=
+             AllocatorSizeAssoc.end();
+    }
+  return false;
+}
+
+static ArrayBoundsInformation::BOUNDSINFOTYPE
+getAllocatedSizeExpr(Expr *E, ASTContext *C, ProgramInfo &Info,
+                     FieldDecl *isField = nullptr) {
+  assert(isAllocatorCall(E) && "The provided expression should be a call to "
+                                "to a known allocator function.");
+  auto &arrBoundsInfo = Info.getArrayBoundsInformation();
+  CallExpr *CE = dyn_cast<CallExpr>(removeAuxillaryCasts(E));
+  std::string funcName = getCalledFunctionName(CE);
+  std::string sizeExpr = "";
+  ArrayBoundsInformation::BOUNDSINFOTYPE previousBoundsInfo;
+  bool isFirstExpr = true;
+  for (auto parmIdx : AllocatorSizeAssoc[funcName]) {
+    Expr *e = CE->getArg(parmIdx);
+    auto currBoundsInfo = arrBoundsInfo.getExprBoundsInfo(isField, e);
+    if (!isFirstExpr) {
+      currBoundsInfo =
+          arrBoundsInfo.combineBoundsInfo(isField, previousBoundsInfo,
+                                          currBoundsInfo, "*");
+      isFirstExpr = false;
+    }
+    previousBoundsInfo = currBoundsInfo;
+  }
+  return previousBoundsInfo;
+
+}
+
+// Check if expression is a simple local variable i.e., ptr = .if yes, return
+// the referenced local variable as the return value of the argument.
+bool isExpressionSimpleLocalVar(Expr *toCheck, VarDecl **targetDecl) {
+  if (DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(toCheck))
+    if (DeclaratorDecl *FD = dyn_cast<DeclaratorDecl>(DRE->getDecl()))
+      if (!dyn_cast<FieldDecl>(FD) && !dyn_cast<ParmVarDecl>(FD))
+        if (VarDecl *VD = dyn_cast<VarDecl>(FD))
+          if (!VD->hasGlobalStorage()) {
+            *targetDecl = VD;
+            return true;
+          }
+  return false;
+}
+
+bool isExpressionStructField(Expr *toCheck, FieldDecl **targetDecl) {
+  if (MemberExpr *DRE = dyn_cast<MemberExpr>(toCheck)) {
+    if (FieldDecl *FD = dyn_cast<FieldDecl>(DRE->getMemberDecl())) {
+      *targetDecl = FD;
+      return true;
+    }
+  }
+  return false;
+}
+
 // This visitor handles the bounds of function local array variables.
 
+void GlobalABVisitor::SetParamHeuristicInfo(LocalVarABVisitor *LAB) {
+  this->ParamInfo = LAB;
+}
+
+bool GlobalABVisitor::IsPotentialLengthVar(ParmVarDecl* PVD) {
+  if (PVD->getType().getTypePtr()->isIntegerType()) {
+    return ParamInfo == nullptr || !ParamInfo->isNonLengthParameter(PVD);
+  }
+  return false;
+}
+
+// This handles the length based heuristics for structure fields.
+bool GlobalABVisitor::VisitRecordDecl(RecordDecl *RD) {
+  // For each of the struct or union types.
+  if (RD->isStruct() || RD->isUnion()) {
+    // Get fields that are identified as arrays and also fields that could be
+    // potential be the length fields
+    std::set<FieldDecl*> potentialLengthFields;
```

Comment:
There should be a space between the type and the *:
`
std::set<FieldDecl *>
`